lib/rust/parser/src/syntax/operator.rs

```diff
@@ -70,110 +70,180 @@ pub fn resolve_operator_precedence_if_non_empty<'s>(
-    let mut precedence_error = None;
+    let mut expression = ExpressionBuilder { nospace, ..default() };
     for item in items {
-        if let syntax::Item::Token(
-                Token { variant: token::Variant::Operator(opr), left_offset, code }) = item {
-            // Item is an operator.
-            if let Some(unsatisified_opr) = unary_operator.take() {
-                output.push(syntax::Tree::unary_opr_app(unsatisified_opr, None).into());
-                prev_type = Some(Ast);
-            }
-            let prev_type = mem::replace(&mut prev_type, Some(Opr));
-            if opr.can_be_binary_infix {
-            } else if opr.can_be_unary_prefix {
-                if prev_type == Some(Ast) {
-                    operator_stack.push(default());
-                }
-                let opr = Token(left_offset, code, opr);
-                unary_operator = Some(opr);
-                continue;
-            } else {
-                precedence_error.get_or_insert_with(|| format!("Precedence of: {:?}", code));
-            };
-            let prec = opr.precedence;
-            let opr = Token(left_offset, code, opr);
-            if prev_type == Some(Opr) && let Some(prev_opr) = operator_stack.last_mut() {
-                // Error. Multiple operators next to each other.
-                prev_opr.push(opr);
-            } else {
-                // Application has the highest precedence.
-                const APP_PREC: usize = std::usize::MAX;
-                while let Some(prev_opr) = operator_stack.last()
-                    && prev_opr.first().map(|opr| opr.precedence).unwrap_or(APP_PREC) >= prec
-                    && let Some(prev_opr) = operator_stack.pop()
-                    && let Some(rhs) = output.pop()
-                {
-                    // Prev operator in the [`operator_stack`] has a higher precedence.
-                    let lhs = output.pop().map(|t| t.to_ast());
-                    if lhs.is_none() { was_section_used = true; }
-                    let ast = syntax::tree::apply_operator(lhs, prev_opr, Some(rhs.to_ast()));
-                    output.push(ast.into());
-                }
-                operator_stack.push(vec![opr]);
-            }
-        } else if let Some(opr) = unary_operator.take() {
-            let rhs = Some(item.to_ast());
-            output.push(syntax::Tree::unary_opr_app(opr, rhs).into());
-            prev_type = Some(Ast);
-        } else if prev_type == Some(Ast) && let Some(lhs) = output.pop() {
+        if let syntax::Item::Token(Token {
+            variant: token::Variant::Operator(opr),
+            left_offset,
+            code,
+        }) = item
+        {
+            expression.operator(Token(left_offset, code, opr));
+        } else {
+            expression.operand(item);
+        }
+    }
+    expression.finish()
+}
+
+
+// === Expression builder ===
+
+/// Stack machine that builds an expression from syntax nodes according to a shift-reduce algorithm.
```

Comment:
put some more info about the algorithm .Maybe a link to shuffling-yard would be useful here too :)